src/contrib/mlir/backend/pass/affine_lowerer.cpp

```diff
@@ -878,22 +908,344 @@ namespace
+
+        NGRAPH_CHECK(!groupsInFilters || groups == filtersShape[0]);
+
+        LoopBuilder::makeAffine(&iv, lb, ub, 1)([&] {
+            // lower/upper bounds on image channel dim and kernels dim
+            auto cLb = iv * channelGroupSize;
+            auto cUb = cLb + channelGroupSize;
+            auto kLb = iv * filtersGroupSize;
+            auto kUb = kLb + filtersGroupSize;
+            lowerConvolution(result,
+                             images,
+                             filters,
+                             strides,
+                             padBelow,
+                             padAbove,
+                             rewriter,
+                             pass,
+                             gConvOp.getLoc(),
+                             cLb,
+                             cUb,
+                             kLb,
+                             kUb,
+                             iv);
+        });
+        rewriter.replaceOp(op, {result});
+        return matchSuccess();
+    }
+    REWRITER(NGReturnOp)
+    {
+        pass.insertDeallocs(rewriter);
+        rewriter.replaceOpWithNewOp<ReturnOp>(op);
+        return matchSuccess();
+    }
+
+#undef REWRITER
+    /// End of pattern matchers
+    template <typename OP>
+    void lowerUnaryElementwise(Operation* op,
+                               ArrayRef<Value*> operands,
+                               PatternRewriter& rewriter,
+                               DialectLoweringPass& pass)
+    {
+        auto loc = cast<OP>(op).getLoc();
+
+        auto result = pass.buildOutputDefs(op, rewriter)[0];
+        NGRAPH_CHECK(result->getType().isa<MemRefType>());
+        // Note that builder's current function is still the original function body.
+        // use getBlock to get the new block instead.
+
+        // get new operands
+        Value* lhs = operands[0];
+
+        ScopedContext scope(rewriter, loc);
+        // Views
+        MemRefView vRes(result), vLHS(lhs);
+        // Index Values
+        IndexedValue iRes(result), iLHS(lhs);
+        // Bounds Index Handles
+        auto lbs = vLHS.getLbs();
+        auto ubs = vLHS.getUbs();
+        // Loop induction vars
+        auto ivs = makeIndexHandles(vLHS.rank());
+        auto pivs = makeHandlePointers(MutableArrayRef<IndexHandle>(ivs));
+        // Steps
+        auto steps = vLHS.getSteps();
+
+        NGRAPH_CHECK(lhs->getType().isa<MemRefType>());
+        Type elemTy = lhs->getType().cast<MemRefType>().getElementType();
+
+        AffineLoopNestBuilder(pivs, lbs, ubs, steps)([&] {
+            ValueHandle val = iLHS(ivs);
+            if (isa<NGNegOp>(op))
+            {
+                ValueHandle zero = createZeroConstant(elemTy);
+                iRes(ivs) = zero - val;
+            }
+            else
+            {
+                NGRAPH_CHECK(false, "Unsupported op");
+            }
+        });
+
+        rewriter.replaceOp(op, {result});
+    }
+
+    template <typename OP>
```

Comment:
We should be able to reorganize the code so that git doesn't think we are moving all this code around. Maybe moving `lowerConvolution` up?